proposals/type-classes.md

```diff
@@ -0,0 +1,182 @@
+
+```kotlin
+Option(1).combine(Option(1)) // Option(2)
+Option("a").combine(Option("b")) // does not compile. Found `Option<A>: Monoid` instance providing `combine` but no `String: Monoid` instance was in scope
+```
+
+We believe the above proposed encoding fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support typeclasses such as Scala where this is done via implicits.
+
+# Typeclasses over type constructors
+
+We recommend if this proposal is accepted that a lightweight version of higher kinds support is included to unveil the true power of typeclasses through the extensions mechanisms
+
+A syntax that would allow for higher kinds in these definitions may look like this:
+
+```kotlin
+typeclass FunctionK<F<_>, G<_>> {
+  fun <A> invoke(fa: F<A>): G<A>
+}
+
+extension Option2List : FunctionK<Option, List> {
+  fun <A> invoke(fa: Option<A>): List<A> =
+    fa.fold({ emptyList() }, { listOf(it) })
+}
+``` 
+
+If Higher Kinds where added along with typeclasses to the lang an alternate definition to the encoding below:
+
+```kotlin
+typeclass Functor {
+    fun Self.map(b: Self): Self
+}
+```
+
+could be provided such as:
+
+```kotlin
+typeclass F<_> : Functor {
+    fun <A, B> map(fa: F<A>, f: (A) -> B): F<B>
+}
+
+extension Option: Functor {
+    fun <A, B> map(fa: Option<A>, f: (A) -> B): Option<B>
+}
+```
+
+Here `F<_>` refers to a type that has a hole on it such as `Option`, `List`, etc. 
+
+A use of this declaration in a polymorphic function would look like:
+
+```kotlin
+fun <F<_> : Functor, A, B> transform(fa: F<A>, f: (A) -> B): F<B> = F.map(fa, f)
+
+transform(Option(1), { it + 1 }) // Option(2)
+transform("", { it + "b" }) // Does not compile: `String` is not type constructor with shape F<_>
+transform(listOf(1), { it + 1 }) // does not compile: No `List<_>: Functor` instance defined in scope.
+```
+
+Once typeclasses extensions are defined over datatypes the compiler could automatically
+add extensions to the target datatypes unless the target datatype already defines a method with the same signature:
+
+```kotlin
+typeclass F<_> : Functor {
+    fun <A, B> map(fa: F<A>, f: (A) -> B): F<B>
+    fun <A, B> Self.Companion.lift(f: (A) -> B): (F<A>) -> F<B>
+}
+
+extension Option: Functor {
+    fun <A, B> Self.map(fa: Option<A>, f: (A) -> B): Option<B> = ... //does not enable `Option(1).map(Option(1)) becase `Option#map` already exists with the same signature as an instance method
+    fun <A, B> lift(f: (A) -> B): (Option<A>) -> Option<B> = ... //enables Option.lift({n: Int -> n.toString() }) because the Option companion does not define `lift` // Option<Int> -> Option<String>
+}
+```
+
+Some of this examples where originally proposed by Roman Elizarov and the Kategory contributors where these features where originally discussed https://kotlinlang.slack.com/archives/C1JMF6UDV/p1506897887000023 
```

Comment:
Small note seeing this link, you may want to copy/paste the snippets if you believe it provides important context because the Kotlin slack is free and only has the most recent 10k messages.

Reply:
@mkobit thanks, I copied most of the snippets from @elizarov into the proposal so I think we are good if that goes away.